### PR TITLE
feat: Datasource autosave improvements

### LINKED ISSUE #4381: [Feature]-[4000]:Autosave for new datasources needs improvements
## Description
In Appsmith we try to autosave everywhere. Today when a user clicks on a new datasource, it creates new named resource even though the user might have just been exploring.
## Solution
We should not save the datasource if there was nothing typed inside the form. It should just disappear because the user was exploring and not actually creating a new datasource.

### PR DESCRIPTION
**Pull request by sneha122**
## Description

In Appsmith we try to autosave everywhere. Today when a user clicks on a new datasource, it creates new named resource even though the user might have just been exploring.

Fixes #4381 

## Type of change

> Please delete options that are not relevant.

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- Manually

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx [hunk 1]
```diff
@@ -507,7 +508,12 @@ class EmbeddedDatasourcePathComponent extends React.Component<
               datasource && "id" in datasource ? datasource.id : undefined
             }
             enable
-            shouldSave={datasource && !("id" in datasource)}
+            shouldSave={
+              (datasource && !("id" in datasource)) ||
+              (datasource &&
+                !!("id" in datasource) &&
```
Comment (AmanAgarwal041) [inline app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx:514]: Can this be replaced by `datasource?.id == TEMP_DATASOURCE_ID`

**Comment by Aishwarya-U-R** [comment]
User should not be able to click if Delete button is disabled:

https://user-images.githubusercontent.com/91450662/199201466-7e4aa99d-35e0-403d-8dd9-585986412e7f.mov

**Comment by Aishwarya-U-R** [comment]
Showing the newly created DS in EntityExplorer also should be avoided since DS is not really created & user attempting to delete it from EE shows 'Resource not found' toast which causes confusion:

https://user-images.githubusercontent.com/91450662/199202320-0c158b87-e3d2-46ce-b2db-2951e56b5321.mov

**Comment by Aishwarya-U-R** [comment]
Save as DS is broken:

https://user-images.githubusercontent.com/91450662/199215968-0fecbf0e-549e-471e-ad3e-4aacebaad2c4.mov

File: app/client/src/pages/common/datasourceAuth/index.tsx [hunk 2]
```diff
@@ -171,10 +184,15 @@ function DatasourceAuth({
     getPluginTypeFromDatasourceId(state, datasourceId),
   );
 
-  // to check if saving during import flow
-  const isReconnectModelOpen: boolean = useSelector(
-    getIsReconnectingDatasourcesModalOpen,
-  );
+  useEffect(() => {
+    if (!!triggerSave && triggerSave) {
```
[the remark this answers is not in the document]
Comment (sneha122) [reply app/client/src/pages/common/datasourceAuth/index.tsx:188]: Yes it does, ill make the necessary changes

File: app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx (hunk 1 above)
Comment (sneha122) [reply app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx:514]: yes will make the necessary changes

File: app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx [hunk 3]
```diff
@@ -67,7 +96,124 @@ const EditDatasourceButton = styled(AdsButton)`
   }
 `;
 
-class DatasourceSaaSEditor extends JSONtoForm<Props> {
+type State = {
```
Comment (AmanAgarwal041) [inline app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx:99]: Can we add comments here to describe about each state ?

File: app/client/src/sagas/QueryPaneSagas.ts [hunk 4]
```diff
@@ -279,16 +287,45 @@ function* handleDatasourceCreatedSaga(actionPayload: ReduxAction<Datasource>) {
...
-    datasourcesEditorIdURL({
-      pageId,
-      datasourceId: actionPayload.payload.id,
-      params: { from: "datasources", ...getQueryParams() },
-    }),
+  const generateCRUDSupportedPlugin: GenerateCRUDEnabledPluginMap = yield select(
+    getGenerateCRUDEnabledPluginMap,
   );
+
+  if (
```
Comment (AmanAgarwal041) [inline app/client/src/sagas/QueryPaneSagas.ts:302]: Can we add a comment here as well to describe about the multiple conditions for a different page change ?

File: app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx (hunk 3 above)
Comment (sneha122) [reply app/client/src/pages/Editor/SaaSEditor/DatasourceForm.tsx:106]: Yes I have made the corresponding changes, can you please check?

File: app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx [hunk 5]
```diff
@@ -381,43 +409,52 @@ class DatasourceRestAPIEditor extends React.Component<
...
     const {
+      datasourceDeleteTrigger,
       datasourceId,
       deleteDatasource,
       hiddenHeader,
       isDeleting,
       isSaving,
     } = this.props;
+
+    const handleDeleteDatasource = (datasourceId: string) => {
```
[the remark this answers is not in the document]
Comment (sneha122) [reply app/client/src/pages/Editor/DataSourceEditor/RestAPIDatasourceForm.tsx]: Yes I have made the corresponding changes, can you please check?

File: app/client/src/sagas/QueryPaneSagas.ts (hunk 4 above)
Comment (sneha122) [reply app/client/src/sagas/QueryPaneSagas.ts:306]: Yes I have made the corresponding changes, can you please check?

**Review by AmanAgarwal041 — DISMISSED** [review]
Looks good to me

**Review by AmanAgarwal041 — DISMISSED** [review]

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/OtherUIFeatures/GlobalSearch_spec.js [hunk 6]
```diff
@@ -169,25 +167,19 @@ describe("GlobalSearch", function() {
       .then((title) => expect(title).includes("Api"));
   });
 
-  it("8. navigatesToGoogleSheetsQuery does not break again: Bug 15012", () => {
+  // since now datasource will only be saved once user clicks on save button explicitly,
+  // updated test so that when user clicks on google sheet and searches for the same datasource, no
+  // results found will be shown
+  it.only("8. navigatesToGoogleSheetsQuery does not break again: Bug 15012", () => {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/OtherUIFeatures/GlobalSearch_spec.js:173]: Can we remove `.only` from here?

File: app/client/cypress/support/commands.js [hunk 7]
```diff
@@ -890,16 +890,17 @@ Cypress.Commands.add("setTinyMceContent", (tinyMceId, content) => {
 
 Cypress.Commands.add("startRoutesForDatasource", () => {
   cy.server();
-  cy.route("PUT", "/api/v1/datasources/*").as("saveDatasource");
+  cy.route("POST", "/api/v1/datasources").as("saveDatasource");
   cy.route("POST", "/api/v1/datasources/test").as("testDatasource");
+  cy.intercept("PUT", "/api/v1/datasources").as("updateDatasource");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/commands.js:895]: Please update `/*` in `updateDatasource` here also!

File: app/client/cypress/support/commands.js [hunk 8]
```diff
@@ -1008,6 +1009,7 @@ Cypress.Commands.add("startServerAndRoutes", () => {
   cy.intercept("GET", "/api/v1/app-templates").as("fetchTemplate");
   cy.intercept("POST", "/api/v1/app-templates/*").as("importTemplate");
   cy.intercept("GET", "/api/v1/app-templates/*").as("getTemplatePages");
+  cy.intercept("PUT", "/api/v1/datasources").as("updateDatasource");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/commands.js:1012]: Please update `/*` in `updateDatasource` here also!

**Review by Aishwarya-U-R — COMMENTED** [review]
Cypress looks good!

**Review by Aishwarya-U-R — APPROVED** [review]

**Pull request merged**